libvirt.go

```diff
@@ -201,19 +216,74 @@ func (l *Libvirt) Events(dom string) (<-chan DomainEvent, error) {
 		return nil, err
 	}
 
-	stream := make(chan *DomainEvent)
+	stream := newStream(constants.QEMUConnectDomainMonitorEventDeregister, constants.ProgramQEMU)
 	l.addStream(cbID, stream)
 	c := make(chan DomainEvent)
+	go func() {
+		// process events
+		for e := range stream.events {
+			c <- *e.(*DomainEvent)
+		}
+	}()
+
+	return c, nil
+}
+
+// LifecycleEvents streams lifecycle events.
+// If a problem is encountered setting up the event monitor connection
+// an error will be returned. Errors encountered during streaming will
+// cause the returned event channel to be closed.
+func (l *Libvirt) LifecycleEvents() (<-chan DomainEventLifecycleMsg, error) {
+	stream, err := l.eventCallbackRegister(int64(DomainEventIDLifecycle))
+	if err != nil {
+		return nil, err
+	}
+
+	c := make(chan DomainEventLifecycleMsg)
 	go func() {
 		// process events
 		for e := range stream {
-			c <- *e
+			c <- e.(*DomainEventCallbackLifecycleMsg).Msg
 		}
 	}()
 
 	return c, nil
 }
 
+// eventCallbackRegister registers a callback for the events of type `eventID`.
+// If `domain` is passed, the callback will be called only for the events of the domain.
+func (l *Libvirt) eventCallbackRegister(eventID int64) (chan event, error) {
```

Comment:
This routine exists in the generated code; it's called `ConnectDomainEventCallbackRegisterAny`: https://github.com/digitalocean/go-libvirt/blob/07f4be593e4b39f221afb121fa7c9c22c6bb36dd/libvirt.gen.go#L13455 It returns a CallbackID, not a stream of events, but this code should use that routine instead.

Reply:
It's indeed a lot better